File: algorithms/189.swift
// https://leetcode.com/problems/rotate-array

class Solution {
    func rotate(inout nums: [Int], _ k: Int) {
        func inplace_reverse(inout arr: [Int], _ start: Int, _ length: Int) {
            var i = start
            var j = i+length-1
            var tmp = 0
            while i < j {
                tmp = arr[i]
                arr[i] = arr[j]
                arr[j] = tmp
                i += 1
                j -= 1
            }
        }

        let k = k % nums.count
        inplace_reverse(&nums, 0, nums.count)
        inplace_reverse(&nums, 0, k)
        inplace_reverse(&nums, k, nums.count-k)
    }
}

File: algorithms/189.py
# https://leetcode.com/problems/rotate-array


class Solution(object):
    def rotate(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: void Do not return anything, modify nums in-place instead.
        """

        k = k % len(nums)

        # Solution - 2
        #
        def inplace_reverse(arr, start, length):
            if length <= 1:
                return
            i = start
            j = start + length - 1
            while i < j:
                arr[i], arr[j] = arr[j], arr[i]
                i += 1
                j -= 1

        inplace_reverse(nums, 0, len(nums))
        inplace_reverse(nums, 0, k)
        inplace_reverse(nums, k, len(nums) - k)

        # Solution - 1
        #
        # while k > 0:
        #     nums.insert(0, nums.pop())
        #     k -= 1
